src/vs/base/browser/performance.ts

```diff
@@ -0,0 +1,231 @@
+		render: EventPhase.Before,
+		selection: EventPhase.Before
+	};
+
+	/**
+	 * Mark the start of the keydown event.
+	 */
+	export function markKeydownStart() {
+		performance.mark('inputlatency/start');
+		performance.mark('keydown/start');
+		state.keydown = EventPhase.InProgress;
+		queueMicrotask(() => markKeydownEnd());
+	}
+
+	/**
+	 * Mark the end of the keydown event.
+	 */
+	function markKeydownEnd() {
+		// Only measure the first render after keyboard input
+		performance.mark('keydown/end');
+		state.keydown = EventPhase.Finished;
+	}
+
+	/**
+	 * Mark the start of the input event.
+	 */
+	export function markInputStart() {
+		performance.mark('input/start');
+		state.input = EventPhase.InProgress;
+	}
+
+	/**
+	 * Mark the end of the input event.
+	 */
+	export function markInputEnd() {
+		queueMicrotask(() => {
+			performance.mark('input/end');
+			state.input = EventPhase.Finished;
+		});
+	}
+
+	/**
+	 * Mark the start of the animation frame performing the rendering.
+	 */
+	export function markRenderStart() {
+		// Render may be triggered during input, but we only measure the following animation frame
+		if (state.keydown === EventPhase.Finished && state.input === EventPhase.Finished && state.render === EventPhase.Before) {
+			// Only measure the first render after keyboard input
+			performance.mark('render/start');
+			state.render = EventPhase.InProgress;
+			queueMicrotask(() => markRenderEnd());
+		}
+	}
+
+	/**
+	 * Mark the end of the animation frame performing the rendering.
+	 *
+	 * An input latency sample is complete when both the textarea selection change event and the
+	 * animation frame performing the rendering has triggered.
+	 */
+	function markRenderEnd() {
+		performance.mark('render/end');
+		state.render = EventPhase.Finished;
+		record();
+	}
+
+	/**
+	 * Mark when the editor textarea selection change event occurs.
+	 *
+	 * An input latency sample is complete when both the textarea selection change event and the
+	 * animation frame performing the rendering has triggered.
+	 */
+	export function markTextareaSelection() {
+		state.selection = EventPhase.Finished;
+		record();
+	}
+
+	/**
+	 * Record the input latency sample if it's ready.
+	 */
+	function record() {
+		// Skip recording this frame if the buffer is full
+		if (measurementsCount >= Constants.bufferLength) {
+			return;
+		}
+		// Selection and render must have finished to record
+		if (state.selection !== EventPhase.Finished || state.render !== EventPhase.Finished) {
+			return;
+		}
+		// Finish the recording, use a timer to ensure that layout/paint is captured. setImmediate
+		// is used if available (Electron) to get slightly more accurate results
+		('setImmediate' in window ? (window as any).setImmediate : setTimeout)(() => {
+			if (state.keydown === EventPhase.Finished && state.input === EventPhase.Finished && state.selection === EventPhase.Finished && state.render === EventPhase.Finished) {
+				performance.mark('inputlatency/end');
```

Comment:
How does this work in cases when a keydown is not followed by an input, like for example when pressing `Backspace` and then sometime later pressing an input producing key like `A`?

Reply:
Is just gets ignored I believe, the `state` stuff was added for this

Reply:
Yes it's the second one, here's a profile of backspace, backspace, backspace, "abc":

![image](https://user-images.githubusercontent.com/2193314/201385401-2940435d-d1c9-4699-862e-73044d43ec9c.png)